docs/contributor/adr/004-layered-architecture.md

```diff
@@ -0,0 +1,77 @@
+# ADR 004 - Adoption of Layered Architecture (3-Tier)
+
+## Status
+
+Accepted
+
+## Context
+
+We need to design a maintainable, testable, and scalable system for [Your Project Name], which interacts with Kubernetes APIs and includes business logic and orchestration of resources.
+
+To achieve this, we have adopted a Layered Architecture pattern with the following three layers:
+1. Controller Layer (Top)
+2. Service Layer (Middle)
+3. Repository Layer (Bottom) 
+
+Each layer has distinct responsibilities and clear dependency rules to ensure separation of concerns and minimize coupling.
+
+## Decision
+
+### Controller Layer
+Responsibility:
+- Serves as the orchestrator for reconciling Kubernetes resources
+- Handles event processing and decision-making around requeue intervals and error handling
+
+Key Points:
+- Depends on Service Layer
+- Uses the results from service calls to make decisions about reconciliation behavior
+- It is the only layer that creates and manages controller-runtime’s ctrl.Result objects
+- Does not perform business logic directly; delegates all such work to services
+
+### Service Layer
+Responsibility:
+- Implements business logic by orchestrating one or more repositories
+- Encapsulates complex workflows and business rules
+
+Key Points:
+- Depends on Repository Layer
+- Must not access Kubernetes APIs directly; instead, uses repositories for any data access
+- Contains pure business logic to ensure it can be tested independently of Kubernetes
+
+### Repository Layer
+Responsibility:
+- Acts as the CRUD (Create, Read, Update, Delete) data layer
+- Directly interacts with the Kubernetes API using the controller-runtime client
+- Responsible for all data persistence and retrieval operations
+
+Key Points:
+- This is the only layer that has direct access to the Kubernetes API
+- Provides a clean abstraction over client calls to isolate API details from the rest of the application
+
+#### Dependency Direction
+![dep dir](./../assets/layered-arch-dep-dir.svg)
```

Comment:
Since it's ADR, it's better based on UML standard, for the arrow, should be a dashed line:  https://www.uml-diagrams.org/dependency.html#:~:text=A%20dependency%20is%20generally%20shown%20as%20a%20dashed,with%20an%20optional%20stereotype%20and%20an%20optional%20name.